services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/SubnetPortsCache.java

```diff
@@ -57,6 +77,70 @@ public synchronized void addSubnetPorts(InternalSubnetPorts internalSubnetPorts)
+        if (networkConfig.getInternalRouterInfos() != null) {
+            Map<String, String> internalSubnetsRouterMap = networkConfig
+                    .getInternalRouterInfos()
+                    .stream()
+                    .filter(routerInfo -> routerInfo.getRouterConfiguration().getSubnetRoutingTables().size() > 0)
+                    .flatMap(routerInfo -> routerInfo.getRouterConfiguration().getSubnetRoutingTables()
+                            .stream()
+                            .map(routingTable -> new InternalSubnetRouterMap(routerInfo.getRouterConfiguration().getId()
+                                    , routingTable.getSubnetId())))
+                    .distinct()
+                    .collect(Collectors.toMap(routerInfo -> routerInfo.getSubnetId(), routerInfo -> routerInfo.getRouterId()));
+            return internalSubnetsRouterMap;
+        }
+        return new HashMap<>();
+    }
+
+    @DurationStatistics
+    public Map<String, InternalSubnetPorts> attacheRouter(Map<String, String> subnetIdRouterIdMap) throws CacheException {
+        Map<String, InternalSubnetPorts> internalSubnetEntityMap = new TreeMap<>();
+        for (Map.Entry<String, String> subnetIdRouterId : subnetIdRouterIdMap.entrySet()) {
+            InternalSubnetPorts internalSubnetPorts = null;
+            try {
+                if (subnetPortsCache.containsKey(subnetIdRouterId.getKey())) {
+                    internalSubnetPorts = subnetPortsCache.get(subnetIdRouterId.getKey());
+                    internalSubnetPorts.setRouterId(subnetIdRouterId.getValue());
+                    internalSubnetEntityMap.put(subnetIdRouterId.getKey(), internalSubnetPorts);
+                }
+            } catch (CacheException e) {
+                e.printStackTrace();
+            }
+        }
+
+        subnetPortsCache.putAll(internalSubnetEntityMap);
+        return internalSubnetEntityMap;
+    }
+
+
+    @DurationStatistics
+    public Map<String, InternalSubnetPorts> getSubnetPorts(NetworkConfiguration networkConfig) throws CacheException {
+        Map<String, String> internalSubnetsRouterMap = getInternalSubnetRouterMap(networkConfig);
+        Map<String, InternalSubnetPorts> internalSubnetPortsMap = new TreeMap<>();
+        for (InternalSubnetEntity subnetEntity : networkConfig.getSubnets()) {
+            InternalSubnetPorts internalSubnetPorts = new InternalSubnetPorts(subnetEntity.getId()
+                    ,subnetEntity.getGatewayPortDetail().getGatewayPortId()
+                    ,subnetEntity.getGatewayIp()
+                    ,subnetEntity.getGatewayPortDetail().getGatewayMacAddress()
+                    ,subnetEntity.getName()
+                    ,subnetEntity.getCidr()
+                    ,subnetEntity.getVpcId()
+                    ,subnetEntity.getTunnelId()
+                    ,subnetEntity.getDhcpEnable()
+                    ,internalSubnetsRouterMap.getOrDefault(subnetEntity.getId(), null));
```

Comment:
This part appears to convert a subnet entity to a internalSubnetPorts object. Could we do it through a conversion function if it could be reused in the future?